Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp

```diff
@@ -195,6 +195,173 @@ void CalleeGroup::compileAsync(VM& vm, AsyncCompilationCallback&& task)
+    // the updates to B but still not have reset its cache of A', which would lead to all kinds of badness.
+    resetInstructionCacheOnAllThreads();
+    WTF::storeStoreFence(); // This probably isn't necessary but it's good to be paranoid.
+
+    m_wasmIndirectCallEntryPoints[functionIndex] = entrypoint;
+
+    if (auto iter = m_jsEntrypointCallees.find(functionIndex); iter != m_jsEntrypointCallees.end())
+        iter->value->setReplacementTarget(entrypoint);
+
+    // FIXME: This does an icache flush for each repatch but we
+    // 1) only need one at the end.
+    // 2) probably don't need one at all because we don't compile wasm on mutator threads so we don't have to worry about cache coherency.
+    for (auto& callsite : callsites) {
+        dataLogLnIf(verbose, "Repatching call at: ", RawPointer(callsite.callLocation.dataLocation()), " to ", RawPointer(entrypoint.taggedPtr()));
+        MacroAssembler::repatchNearCall(callsite.callLocation, callsite.target);
+    }
+}
+
+void CalleeGroup::reportCallees(const AbstractLocker&, JITCallee* caller, const FixedBitVector& callees)
+{
+#if ASSERT_ENABLED
+    for (const auto& call : caller->wasmToWasmCallsites()) {
+        if (call.functionIndexSpace < functionImportCount())
+            continue;
+        ASSERT(const_cast<FixedBitVector&>(callees).test(toCodeIndex(call.functionIndexSpace)));
+    }
+#endif
+    auto callerIndex = toCodeIndex(caller->index());
+    ASSERT_WITH_MESSAGE(callees.size() == FixedBitVector(m_calleeCount).size(), "Make sure we're not indexing callees with the space index");
+    for (size_t calleeIndex : callees) {
+        // dataLogLn(callerIndex, " reported as now calling ", calleeIndex);
+        m_callers[calleeIndex].testAndSet(callerIndex);
+    }
+}
+#endif
+
+TriState CalleeGroup::calleeIsReferenced(const AbstractLocker&, Wasm::Callee* callee) const
+{
+    switch (callee->compilationMode()) {
+    case CompilationMode::LLIntMode:
+    case CompilationMode::IPIntMode:
+        return TriState::True;
+    case CompilationMode::BBQMode: {
+        FunctionCodeIndex index = toCodeIndex(callee->index());
+        if (m_bbqCallees.at(index).isWeak())
+            return m_bbqCallees.at(index).get() ? TriState::Indeterminate : TriState::False;
+        return triState(m_bbqCallees.at(index).ptr());
+    }
+    case CompilationMode::OMGMode:
+        return triState(m_omgCallees.at(toCodeIndex(callee->index())).get());
+    case CompilationMode::OMGForOSREntryMode: {
+        FunctionCodeIndex index = toCodeIndex(callee->index());
+        if (m_osrEntryCallees.get(index).get()) {
+            // The BBQCallee really owns the OSREntryCallee so as long as that's around the OSREntryCallee is referenced.
+            if (m_bbqCallees.at(index).get())
+                return TriState::True;
+            return TriState::Indeterminate;
+        }
+        return TriState::False;
+    }
```

Comment:
Also `m_bbqCallees`, `m_omgCallees`, `m_osrEntryCallees` all have conditional existence.

Reply:
Ah, sorry, I tried to cover these cases but it's easy to forget...